Redefine signal for architecture mips

sunny868 opened the pull request:
define SIGSTOP 19 is just ok for architecture x86、arm、s390、powerpc and so on.
 but others architecture has different values.
 For architecture mips,you can see kernel source code arch/mips/include/uapi/asm/signal.h

sunny868 committed:
Define SIGSTOP 23 for architecture mips

 define SIGSTOP 19 is just ok for architecture x86、arm、s390、powerpc and so on.
 but others architecture has different values.
 For architecture mips,you can see kernel source code arch/mips/include/uapi/asm/signal.h

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=40299) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=40299) before we can accept your contribution.<br/><br/>:x: sunny868 [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=40299)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=40299) it.</sub>

sunny868 commented:
I'm not sure it's the best way to redefine SIGSTOP for architecture mips in Interop.Kill.cs. Maybe we can redefine SIGSTOP in pal_process.c : 

` _int32_t SystemNative_Kill(int32_t pid, int32_t signal)                         `
` {          `
`      #ifdef __TARGET_MIPS64_`
`          if( signal == 19)`
`                    signal = SIGSTOP;`
`         #endif                                                         `
`    return kill(pid, signal);                                                   `
` }`

sunny868 commented:
The 4 failing has nothing to do  with this PR.

am11 commented on src/libraries/System.Diagnostics.Process/src/System.Diagnostics.Process.csproj:
```diff
@@ -10,6 +10,9 @@
     <GeneratePlatformNotSupportedAssemblyMessage Condition="'$(TargetsAnyOS)' == 'true'">SR.Process_PlatformNotSupported</GeneratePlatformNotSupportedAssemblyMessage>
   </PropertyGroup>
+   <PropertyGroup Condition="'$(Platform)' == 'mips64'">
+     <DefineConstants>MIPS64;$(DefineConstants)</DefineConstants>
```
Per the conversion used by others architectures:

```suggestion
     <DefineConstants>TARGET_MIPS64;$(DefineConstants)</DefineConstants>
```

sunny868 replied:
Thanks!

jkotas commented on src/libraries/System.Diagnostics.Process/src/System.Diagnostics.Process.csproj:
```diff
@@ -10,6 +10,9 @@
   <PropertyGroup>
     <GeneratePlatformNotSupportedAssemblyMessage Condition="'$(TargetsAnyOS)' == 'true'">SR.Process_PlatformNotSupported</GeneratePlatformNotSupportedAssemblyMessage>
   </PropertyGroup>
+   <PropertyGroup Condition="'$(Platform)' == 'mips64'">
```
src/libraries csharp files are built as architecture-neutral. Switching on architecture here is not going to work well with the rest of the build.

The better way to fix this is to remap the signal numbers under https://github.com/dotnet/runtime/tree/master/src/libraries/Native

sunny868 committed:
Redefine signal for architecture mips

 define SIGSTOP 19 is just ok for architecture x86、arm、s390、powerpc and so on.
 but others architecture has different values.
 For architecture mips,you can see kernel source code arch/mips/include/uapi/asm/signal.h

 Co-Authored-By: hev <wangrui@loongson.cn>

jkotas commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.Kill.cs:
```diff
@@ -12,7 +12,12 @@ internal enum Signals : int
         {
             None = 0,
             SIGKILL = 9,
+#if TARGET_MIPS64
```
Should this commit be reverted now?

sunny868 replied:
yes

jkotas commented on src/libraries/Native/Unix/System.Native/pal_process.c:
```diff
@@ -653,6 +653,47 @@ int32_t SystemNative_SetRLimit(RLimitResources resourceType, const RLimit* limit
 
 int32_t SystemNative_Kill(int32_t pid, int32_t signal)
 {
+#if defined(TARGET_MIPS64)
```
I would be better to define `PAL_...` specific enum in pal_process.h (similar to other enums that exist there already). This enum should have just the few signals that we actually need, and this remapping can then be done unconditionally, without hardcoding of magic numbers.

sunny868 replied:
I'm worried that more signals will be added in Interop.Kill.cs (internal enum Signals:int{..}) later.

jkotas replied:
If that happens, both places will need to be updated. It is same with all other PAL enums.

sunny868 replied:
ok

sunny868 committed:
Revert " Define SIGSTOP 23 for architecture mips"

This reverts commit fcc2d46e45804cfe1d39e9b9db86260abfb8be65.

sunny868 committed:
Redefine signal for architecture mips

define SIGSTOP 19 is just ok for architecture x86、arm、s390、powerpc and so on.
but others architecture has different values.
For architecture mips,you can see kernel source code arch/mips/include/uapi/asm/signal.h

Co-Authored-By: hev <wangrui@loongson.cn>

jkotas commented on src/libraries/Native/Unix/System.Native/pal_process.c:
```diff
@@ -653,6 +653,12 @@ int32_t SystemNative_SetRLimit(RLimitResources resourceType, const RLimit* limit
+    switch (signal)
+    {
+        case PAL_SIGSTOP:
+             signal = SIGSTOP;
+    }
```
```suggestion
             signal = SIGSTOP;
             break;
    }
```

jkotas commented on src/libraries/Native/Unix/System.Native/pal_process.c:
```diff
@@ -653,6 +653,12 @@ int32_t SystemNative_SetRLimit(RLimitResources resourceType, const RLimit* limit
 
 int32_t SystemNative_Kill(int32_t pid, int32_t signal)
```
Maybe this can even be:

```
    switch (signal)
    {
        case PAL_SIGKILL:
             signal = SIGKILL;
             break;

        case PAL_SIGSTOP:
             signal = SIGSTOP;
             break;

        default:
             assert_msg(false, "Unknown signal");
             break;
    }
```

And delete `c_static_assert(PAL_SIGKILL == SIGKILL);` at the top of the file. I have verified that clang optimizes the switch out when the remap-signal numbers  match.

sunny868 replied:
Thanks!

sunny868 commented:
Maybe The 11 failing has nothing to do with this PR.

jkotas reviewed (approved):
Thanks

safern commented:
@jkotas it seems like this PR was merged with System.Diagnostics.Process test failures hitting the assert that was added as part of this PR:

```
/root/helix/work/workitem /root/helix/work/workitem
  Discovering: System.Diagnostics.Process.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Diagnostics.Process.Tests (found 235 of 291 test cases)
  Starting:    System.Diagnostics.Process.Tests (parallel test collections = on, max threads = 2)
    System.Diagnostics.Tests.ProcessStartInfoTests.ShellExecute_Nano_Fails_Start [SKIP]
      Condition(s) not met: "IsWindowsNanoServer"
Linux
Linux
Linux
Linux
None of the following programs were installed on this machine: xdg-open,gnome-open,kfmclient.
/__w/1/s/src/libraries/Native/Unix/System.Native/pal_process.c (665): error 0: Unknown signal (false failed)
Assertion failed: false && "assert_msg failed" (/__w/1/s/src/libraries/Native/Unix/System.Native/pal_process.c: SystemNative_Kill: 665)
./RunTests.sh: line 161:    24 Aborted                 (core dumped) "$RUNTIME_PATH/dotnet" exec --runtimeconfig System.Diagnostics.Process.Tests.runtimeconfig.json --depsfile System.Diagnostics.Process.Tests.deps.json xunit.console.dll System.Diagnostics.Process.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing $RSP_FILE
/root/helix/work/workitem
```

safern commented:
I put up a revert here: https://github.com/dotnet/runtime/pull/40470

jkotas commented:
I am sorry. Once I have seen the known problem "CannotRemoveAttribute : Attribute 'System.ComponentModel.TypeConverterAttribute' exists on 'System.ComponentModel.MarshalByValueComponent' " , I have stopped looking through all the failures.

jkotas commented:
@sunny868 Could you please resubmit with the fix?

safern commented:
> I am sorry. Once I have seen the known problem "CannotRemoveAttribute : Attribute 'System.ComponentModel.TypeConverterAttribute' exists on 'System.ComponentModel.MarshalByValueComponent' " , I have stopped looking through all the failures.

Yeah it is understandable when CI is on the floor because of other things, that things like this could slip through. I'm working on fixing the ApiCompat problem.

sunny868 commented:
> @sunny868 Could you please resubmit with the fix?

Sorry, I I forgot to reset value "Interop.Sys.Signals.None"  to PAL_NONE . I'll resubmit it later.

sunny868 commented:
@jkotas I have resubmit it ,please review : https://github.com/dotnet/runtime/pull/40513
